src/asm/aarch64/cdef.rs

```diff
@@ -0,0 +1,407 @@
+// Copyright (c) 2020, The rav1e contributors. All rights reserved
+//
+// This source code is subject to the terms of the BSD 2 Clause License and
+// the Alliance for Open Media Patent License 1.0. If the BSD 2 Clause License
+// was not distributed with this source code in the LICENSE file, you can
+// obtain it at www.aomedia.org/license/software. If the Alliance for Open
+// Media Patent License 1.0 was not distributed with this source code in the
+// PATENTS file, you can obtain it at www.aomedia.org/license/patent.
+
+use crate::cdef::*;
+use crate::cpu_features::CpuFeatureLevel;
+use crate::frame::*;
+use crate::tiling::PlaneRegionMut;
+use crate::util::*;
+
+type CdefPaddingFn = unsafe extern fn(
+  tmp: *mut u16,
+  src: *const u8,
+  src_stride: isize,
+  left: *const [u8; 2],
+  top: *const u8,
+  h: i32,
+  edges: isize,
+);
+
+type CdefPaddingHBDFn = unsafe extern fn(
+  tmp: *mut u16,
+  src: *const u16,
+  src_stride: isize,
+  left: *const [u16; 2],
+  top: *const u16,
+  h: i32,
+  edges: isize,
+);
+
+type CdefFilterFn = unsafe extern fn(
+  dst: *mut u8,
+  dst_stride: isize,
+  tmp: *const u16,
+  pri_strength: i32,
+  sec_strength: i32,
+  dir: i32,
+  damping: i32,
+  h: i32,
+  edges: isize,
+);
+
+type CdefFilterHBDFn = unsafe extern fn(
+  dst: *mut u16,
+  dst_stride: isize,
+  tmp: *const u16,
+  pri_strength: i32,
+  sec_strength: i32,
+  dir: i32,
+  damping: i32,
+  h: i32,
+  edges: isize,
+  bitdepth_max: i32,
+);
+
+#[inline(always)]
+const fn decimate_index(xdec: usize, ydec: usize) -> usize {
+  ((ydec << 1) | xdec) & 3
+}
+
+pub(crate) unsafe fn cdef_filter_block<T: Pixel>(
+  dst: &mut PlaneRegionMut<'_, T>, src: *const T, src_stride: isize,
+  pri_strength: i32, sec_strength: i32, dir: usize, damping: i32,
+  bit_depth: usize, xdec: usize, ydec: usize, edges: u8, cpu: CpuFeatureLevel,
+) {
+  let call_rust = |dst: &mut PlaneRegionMut<T>| {
+    rust::cdef_filter_block(
+      dst,
+      src,
+      src_stride,
+      pri_strength,
+      sec_strength,
+      dir,
+      damping,
+      bit_depth,
+      xdec,
+      ydec,
+      edges,
+      cpu,
+    );
+  };
+
+  #[cfg(feature = "check_asm")]
+  let ref_dst = {
+    let mut copy = dst.scratch_copy();
+    call_rust(&mut copy.as_region_mut());
+    copy
+  };
+  match T::type_enum() {
+    PixelType::U8 => {
+      match (
+        CDEF_PAD_FNS[cpu.as_index()][decimate_index(xdec, ydec)],
+        CDEF_FILTER_FNS[cpu.as_index()][decimate_index(xdec, ydec)],
+      ) {
+        (Some(pad), Some(func)) => {
```

Comment:
pad and filter are always implemented together, ideally we could make a single dispatch instead of two.